woocommerce/woocommerce-ios
# [Mobile Payments] Create Payment Intent

Closes #3748 and #3749 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️

🔴🔴 Testing this PR requires a hardware reader and setting up a store for testing🔴🔴

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by implementing logic in Yosemite and Hardware to create a Payment Intent on device. 

The significance of a Payment Intent is, hopefully, explained in [the README accompanying the PR](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/HARDWARE.md).

## Changes
* Implemented `createPaymentIntent` in StripeCardReaderService to implement just enough functionality to attack the Stripe Terminal SDK and create a PaymentIntent. This does not include error handling for now. #3741 and #3747 will cover error handling.
* Implements support in `CardPresentPaymentAction` and `CardPresentPaymentStore` for a new high-level action to collect a payment. The reason why the high level action exposed to the app is "collect a payment" instead of just "create an intent" is covered [in the README that is part of the PR diff](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/HARDWARE.md).
* Adds a variable to `OrderDetailsViewModel` that exposes if an order is eligible for card present payments. At this point, that variable relies only on the feature flag, but eventually this is where we would check if the user is a WCPay user and the order is eligible.
* Adds a button to the toolbar of the OrderDetailsViewController only for testing purposes. This button is available in all orders for now, and triggers the "collect payment" high level action exposed by Yosemite The button will be removed and implemented properly in #3826 

## How to test
Testing is complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

The process to test this PR requires interacting with the UI while running the app on a device, and also looking at the Xcode console for some logs.
1. Checkout the branch. Run `bundle exec pod install`. 
2. Build and run on a device. 
3. Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
4. Log in to an account that matches a store that has been configured for testing.
5. Navigate to Settings in the app. (Tap the gear button in the top right)
6. In the "Store Settings" section, tap "Manage Card Reader".
7. Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
8. Tap "Connect card reader"
9. Wait until the app suggests one reader to connect to.
10. Tap "Connect to reader"
11. Make sure the reader light turns solid blue.
12. Navigate to an order (any order).
13. Tap the dollar sign in the top right corner
14. Check the console in Xcode. There should be a log similar to this:
```
==== Log for testing purposes. payment intent collected 
PaymentIntent(id: "pi_1IVHva2HDOoC1bTcjSAC5BAc", status: Hardware.PaymentIntentStatus.requiresPaymentMethod, created: 2021-03-15 15:01:54 +0000, amount: 100, currency: "usd", metadata: Optional([:]), charges: [])
//// payment intent collected 
```

The status should indicate that the PaymentIntent still requires a payment method.

Step by step screenshots:
Step 1:
<img src="https://user-images.githubusercontent.com/2722505/111180448-6ebca980-8583-11eb-8852-a72bba1a72f8.PNG" width="350"/>

Step 2:
<img src="https://user-images.githubusercontent.com/2722505/111180466-72503080-8583-11eb-8358-1bec5676ae5c.PNG" width="350"/>

Step 3:
<img src="https://user-images.githubusercontent.com/2722505/111180457-6fedd680-8583-11eb-860c-29bd69c96de3.PNG" width="350"/>

Step 4:
<img src="https://user-images.githubusercontent.com/2722505/111180453-6f554000-8583-11eb-907f-a7476b0dce89.PNG" width="350"/>

Step 5:
<img src="https://user-images.githubusercontent.com/2722505/111180461-70866d00-8583-11eb-924d-03cd174466b3.PNG" width="350"/>

Step 6:
<img src="https://user-images.githubusercontent.com/2722505/111180463-71b79a00-8583-11eb-96cf-c56e678acb13.PNG" width="350"/>

Step 7:
<img src="https://user-images.githubusercontent.com/2722505/111180464-71b79a00-8583-11eb-8672-2b0dc9e7a762.PNG" width="350"/>


Update release notes:
